third_party/move/move-vm/runtime/src/tracing.rs

```diff
@@ -2,82 +2,120 @@
 // Copyright (c) The Move Contributors
 // SPDX-License-Identifier: Apache-2.0
 
-#[cfg(feature = "debugging")]
-use crate::debug::DebugContext;
-#[cfg(feature = "debugging")]
-use crate::{interpreter::InterpreterDebugInterface, loader::LoadedFunction, RuntimeEnvironment};
-#[cfg(feature = "debugging")]
-use move_vm_types::instr::Instruction;
-#[cfg(feature = "debugging")]
-use ::{
-    move_vm_types::values::Locals,
-    once_cell::sync::Lazy,
-    std::{
-        env,
-        fs::{File, OpenOptions},
-        io::Write,
-        sync::Mutex,
+use crate::{
+    debug::DebugContext, interpreter::InterpreterDebugInterface, loader::LoadedFunction,
+    RuntimeEnvironment,
+};
+use arc_swap::ArcSwap;
+use move_vm_types::{instr::Instruction, values::Locals};
+use std::{
+    env,
+    fs::{File, OpenOptions},
+    io::{BufWriter, Write},
+    path::{Path, PathBuf},
+    sync::{
+        atomic::{AtomicBool, Ordering},
+        Arc, Mutex, OnceLock,
     },
 };
 
-#[cfg(feature = "debugging")]
 const MOVE_VM_TRACING_ENV_VAR_NAME: &str = "MOVE_VM_TRACE";
-
-#[cfg(feature = "debugging")]
 const MOVE_VM_STEPPING_ENV_VAR_NAME: &str = "MOVE_VM_STEP";
 
-#[cfg(feature = "debugging")]
-static FILE_PATH: Lazy<String> = Lazy::new(|| {
-    env::var(MOVE_VM_TRACING_ENV_VAR_NAME).unwrap_or_else(|_| "move_vm_trace.trace".to_string())
-});
+static FILE_PATH: OnceLock<ArcSwap<String>> = OnceLock::new();
+static TRACING_ENABLED: OnceLock<AtomicBool> = OnceLock::new();
+static DEBUGGING_ENABLED: OnceLock<AtomicBool> = OnceLock::new();
+static LOGGING_FILE_WRITER: OnceLock<Mutex<BufWriter<File>>> = OnceLock::new();
+static DEBUG_CONTEXT: OnceLock<Mutex<DebugContext>> = OnceLock::new();
 
-#[cfg(feature = "debugging")]
-pub static TRACING_ENABLED: Lazy<bool> =
-    Lazy::new(|| env::var(MOVE_VM_TRACING_ENV_VAR_NAME).is_ok());
+/// Turn tracing on or off, saving info to the path.
+pub fn enable_tracing(path_opt: Option<&str>) {
```

Comment:
We never call this concurrently, right?

If so, I think it is better to use guard patter here, returning a guard and then on drop, allowing to reset the file path value? Basically:

```
let g = enable_tracingSome(("..."));
// you concurrent calls to executor
drop(g)

// here we can reset?
```
just using a wrapper around RWLockGuard is enough?